File: backend/app/modules/order/schemas/delivery_schema.py
from datetime import UTC, datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, model_validator

from app.modules.order.models.delivery_details import DeliveryTimingStatus


class DeliverySchema(BaseModel):
    receiptent_name: str
    delivery_address: str
    delivery_timing: DeliveryTimingStatus = DeliveryTimingStatus.ASAP
    delivery_note: str | None = None
    scheduled_time: datetime | None = None
    receiptent_phone: str
    delivery_landmark: str

    @model_validator(mode="after")
    def verify_scheduled_date_for_schedule(cls, values):
        is_scheduled = values.delivery_timing == DeliveryTimingStatus.SCHEDULED
        if is_scheduled and values.scheduled_time is None:
            raise ValueError("Scheduled time is required for scheduled delivery")
        if is_scheduled and values.scheduled_time:
            now = datetime.now(UTC)
            if now > values.scheduled_time:
                raise ValueError("Scheduled date is in past")
        return values


class DeliveryReadBasic(BaseModel):
    receiptent_name: str
    delivery_address: str
    delivery_timing: DeliveryTimingStatus
    delivery_note: str | None = None
    scheduled_time: datetime | None = None
    receiptent_phone: str
    delivery_landmark: str

    model_config = ConfigDict(from_attributes=True)


class DeliveryCreate(DeliveryReadBasic):
    order_id: UUID

File: frontend/src/schemas/order/delivery_schema.ts
import z from "zod";

export const DeliveryTimingStatus = z.enum(["asap", "scheduled"]);

export const deliverySchema = z
  .object({
    receiptent_name: z.string().min(1, "Recipient name is required"),

    delivery_address: z.string().min(1, "Delivery address is required"),

    delivery_timing: DeliveryTimingStatus.default("asap"),

    scheduled_time: z.coerce.date().nullable().optional(),

    receiptent_phone: z
      .string()
      .regex(/^(97|98)\d{8}$/, "Enter a valid Nepali mobile number"),

    delivery_landmark: z.string().min(1, "Delivery landmark is required"),

    delivery_note: z.string().optional(),
  })
  .refine(
    (data) =>
      data.delivery_timing !== "scheduled" || data.scheduled_time != null,
    {
      message: "Scheduled time is required for scheduled delivery",
      path: ["scheduled_time"],
    },
  )
  .refine(
    (data) =>
      data.delivery_timing !== "scheduled" ||
      !data.scheduled_time ||
      data.scheduled_time > new Date(),
    {
      message: "Scheduled date is in the past",
      path: ["scheduled_time"],
    },
  );

export type DeliverySchemaType = z.infer<typeof deliverySchema>;
